Sprint 7 Pull Request

icydtw opened the pull request:
Конец 7-го спринта: тестирование, MVC -> MVP.

icydtw committed:
Небольшие изменения.

icydtw committed:
Добавили тестирование.

icydtw committed:
Внесение изменений перед рефакторингом (MVC -> MVP).

icydtw committed:
Перенос бизнес-логики из контроллера в презентер (1).

icydtw committed:
Переносим бизнес-логику (2).

icydtw committed:
Перенос бизнес-логики (3, 7 шаг)

icydtw committed:
Перенесли всю бизнес-логику в презентер!

icydtw committed:
Сделали тест для Presenter().convert().

icydtw committed:
Небольшие исправления.

icydtw committed:
Мелкие исправления по всему проекту.

icydtw committed:
Delete MQUITests directory

icydtw committed:
Delete MoviesLoaderTests directory

ABashkirova commented:
@icydtw привет, Илья! Проект не компилируется, не хватает файла GameRecord, надо бы запушить

ABashkirova commented on MovieQuiz/Models/MostPopularMovies.swift:
```diff
@@ -27,10 +26,11 @@ struct MostPopularMovie: Codable {
     }
     
-    private enum CodingKeys: String, CodingKey {
+    private enum CodingKeys : String, CodingKey {
```
Обычно двоеточие так не ставят при объявлении протоколов :) 
```suggestion
    private enum CodingKeys: String, CodingKey {
```

ABashkirova commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,127 @@
+import UIKit
+
+final class MovieQuizPresenter: QuestionFactoryDelegate {
+    private let statisticService: StatisticService!
```
здесь нет смысла использовать неявное разворачивание(implicitly unwrapping) потому что в инициализаторе  гарантировано будет установлено значение 
```suggestion
    private let statisticService: StatisticService
```

Во вьюконтроллере так было нельзя, так как присваивание происходило уже после инициализации объекта (во viewDidLoad)

С фабрикой тоже самое

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,150 +1,94 @@
+final class MovieQuizViewController: UIViewController, MovieQuizViewControllerProtocol {
     @IBOutlet private var imageView: UIImageView!
-    @IBOutlet private var questionLabel: UILabel!
+    @IBOutlet private var textLabel: UILabel!
```
кажется до этого название было лучше :)

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,150 +1,94 @@
-        showLoadingIndicator()
+        presenter = MovieQuizPresenter(viewController: self)
+        imageView.layer.cornerRadius = 20
         statisticService = StatisticServiceImplementation()
```
а эта штука переехала же в презентер?

ABashkirova commented on MovieQuiz/Services/NetworkClient.swift:
```diff
@@ -1,32 +1,29 @@
             // Проверяем, что нам пришёл успешный код ответа
             if let response = response as? HTTPURLResponse,
-                response.statusCode < 200 || response.statusCode >= 300 {
+                response.statusCode < 200 && response.statusCode >= 300 {
```
а почему исправили?

ABashkirova commented on MovieQuiz/Services/QuestionFactory.swift:
```diff
@@ -1,75 +1,60 @@
     private let moviesLoader: MoviesLoading
-    weak var delegate: QuestionFactoryDelegate?
-    private var movies: [MostPopularMovie] = []
+    private let delegate: QuestionFactoryDelegate
```
Здесь появится retain cycle :( все еще слабая ссылка нужна

icydtw commented:
> @icydtw привет, Илья! Проект не компилируется, не хватает файла GameRecord, надо бы запушить

Хорошо, сейчас гляну, что там.

ABashkirova commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,63 @@
+
+    func testExample() throws {
+        let app = XCUIApplication()
+        app.launch()
```
Этот тест пользы не дает - удаляем

icydtw committed:
Пушим GameRecord.

ABashkirova commented on MovieQuizUITests/MovieQuizUITests.swift:
```diff
@@ -0,0 +1,63 @@
+        let indexLabel = app.staticTexts["Index"]
+        XCTAssertTrue(indexLabel.label == "1/10")
+    }
+}
```
Здесь ожидается, что реализуете 2 разных теста

ABashkirova commented on MovieQuizUITests/MovieQuizUITestsLaunchTests.swift:
```diff
@@ -0,0 +1,32 @@
+
+import XCTest
+
+final class MovieQuizUITestsLaunchTests: XCTestCase {
```
Этот тестовый набор тоже не несет пользы - избавляемся

ABashkirova commented on MovieQuizUITests/UITests.swift:
```diff
@@ -0,0 +1,65 @@
+
+@testable import MovieQuiz
+
+final class UITests: XCTestCase {
```
Два одинаковых тестовых класса?

icydtw commented:
> @icydtw привет, Илья! Проект не компилируется, не хватает файла GameRecord, надо бы запушить

Проверьте, вроде получилось. @ABashkirova

ABashkirova commented:
@icydtw давай порефакторим лишнее и разобьем на два теста, как просят в задаче. 

Это все, тесты рабочие, рефакторинг верный. 
Есть две ошибки из предыдущего спринта: retain cycle и обработка кода ответа от сервера - тоже надо поправить.

icydtw committed:
Update MovieQuiz/Models/MostPopularMovies.swift

Co-authored-by: Alexandra Bashkirova <s.bshkrva@gmail.com>

icydtw committed:
Update MovieQuiz/Presentation/MovieQuizPresenter.swift

Co-authored-by: Alexandra Bashkirova <s.bshkrva@gmail.com>

icydtw committed:
Убрал statisticService из контроллера.

icydtw committed:
Поправил обработку кода ответа от сервера.

icydtw committed:
Удалил ненужный тестовый тест.

icydtw committed:
Избавился от retain cycle.

icydtw committed:
Разбил testResult на два: testResult() и testHide().

icydtw committed:
Удалил лишнее...

icydtw commented:
@ABashkirova, прошёлся по замечаниям, всё понятно, вроде бы всё поправил (просто последний спринт был очень сумбурным для меня и делался очень быстро), теперь приличнее выглядит. Посмотрите пожалуйста, а то у нас дедлайн сегодня... 🙏🏻

ABashkirova commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,127 @@
+
+final class MovieQuizPresenter: QuestionFactoryDelegate {
+    private let statisticService: StatisticService
+    weak var viewController: MovieQuizViewControllerProtocol?
```
а почему публичное?

ABashkirova commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,127 @@
+    }
+    
+    func didAnswer(isCorrectAnswer: Bool) {
+        if (isCorrectAnswer) {
```
В свифте можно опустить скобки
```suggestion
        if isCorrectAnswer {
```

ABashkirova commented on MovieQuiz/Helpers/UIColor+Extensions.swift:
```diff
@@ -1,4 +1,7 @@
 
-extension UIColor { }
+extension UIColor {
+    static var ypGreen: UIColor { UIColor.green }
+    static var ypRed: UIColor { UIColor.red }
```
Не очень поняла как ваша работа дошла до седьмого спринта с этими цветами :( Приношу извинения, похоже пропустили. Также отсутствуют шрифты отсутствуют в проекте 


По макетам цвета у фона, рамок - другие. Надо поправить
<img width="505" alt="Снимок экрана 2022-11-17 в 20 59 40" src="https://user-images.githubusercontent.com/20269264/202509691-483c1576-9f3b-4819-b00b-dfd7181faafa.png">

ABashkirova commented on MovieQuiz/Presentation/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,127 @@
+        
+        let bestGame = statisticService.bestGame
+        
+        let totalPlaysCountLine = "Количество сыгранных квизов: \(statisticService.gamesCount)"
+        let currentGameResultLine = "Ваш результат: \(correctAnswers)\\\(questionsAmount)"
+        let bestGameInfoLine = "Рекорд: \(bestGame.correct)\\\(bestGame.total)"
+        + " (\(bestGame.date.dateTimeString))"
```
слеши в другую сторону 
<img width="505" alt="Снимок экрана 2022-11-17 в 21 03 04" src="https://user-images.githubusercontent.com/20269264/202510371-c581184e-0744-4006-9e99-cbab9113e223.png">

icydtw committed:
Немного поправил уведомлении при проблеме подключения.

ABashkirova reviewed (changes requested):
>  Посмотрите пожалуйста, а то у нас дедлайн сегодня... 🙏🏻

Работа уже отправлена, в дедлайн вы уложились.
Я предлагаю немного выдохнуть и спокойно привести проект корректному виду. 

Это последний спринт где ты будешь работать над этим приложением, дальше он становится частью твоего портфолио. И к сожалению, проект реализован не полностью: есть существенные отличия от макета. Для будущего работодателя это может стать красным флажком. Поэтому я настаиваю на еще одной итерации исправлений. 

Нужно проверить:
- launch screen должен иметь изображение, заполняющее полностью экран, без деформаций 
- добавлены и включены в таргет проекта шрифты
- цвет фона, цвета рамок - удовлетворяют макетам
- будет здорово, если также будет исправлена верстка главного стека: по макету, контент с низу крепится к safe area с отступов в ноль. К сожалению, добавление этого констрейнта ломает отображение вопроса, он перестанет влезать. Здесь я предлагаю просто уменьшить отступы от его контейнера 5-5 верх них, 10-10 право-лево
- Обращу внимание, что пользователь может случайно прожать два раза кнопку ответа или даже много и быстро понажимать на кнопки.  Это может привести к неверному увеличению счетчика, или лишнего вызова алерта с результатом. Принято такое поведение блокировать на время выполнения операции запроса следующего вопроса.
Можно использовать свойство [isEnabled](https://developer.apple.com/documentation/uikit/uicontrol/1618217-isenabled). До начала смены вопроса "выключать" кнопку, после отображения свежих данных "включать"


Полезный материалы по код стайлу 

- [Подробное описание codestyle](https://google.github.io/swift/) от Google
- [От Ray Wanderlich](https://github.com/raywenderlich/swift-style-guide)
- [На русском от команды tutu](https://github.com/tutu-ru/swift-style-guide)

icydtw committed:
Вернул изображение в лонч-скрин.

icydtw committed:
1. Добавил в лонч-скрин изображение
2. Добавил шрифты и применил на экране
3. Настроил верные цвета
4. Прикрепил Stack View со всеми элементами к Safe Area
5. Исправил баг с многократным нажатием на кнопки
6. + Небольшие исправления в коде

icydtw commented:
@ABashkirova, Вы уж извините меня за назойливость и просьбы посмотреть снова и снова, но можете глянуть теперь? Чтобы я на последнюю 3-ю попытку не рисковал лишний раз отправлять. 😄 Я многое поправил.

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -5,6 +5,8 @@ final class MovieQuizViewController: UIViewController, MovieQuizViewControllerPr
     @IBOutlet private var questionLabel: UILabel!
     @IBOutlet private var counterLabel: UILabel!
     @IBOutlet private var activityIndicator: UIActivityIndicatorView!
+    @IBOutlet weak var yesButtonView: UIButton!
+    @IBOutlet weak var noButtonView: UIButton!
```
private)

ABashkirova commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -20,10 +22,12 @@ final class MovieQuizViewController: UIViewController, MovieQuizViewControllerPr
     // MARK: - Actions
     
     @IBAction private func yesButtonClicked(_ sender: UIButton) {
+        yesButtonView.isEnabled = false
```
По-хорошему обе кнопки блочить

ABashkirova commented on MovieQuiz/Resources/Info.plist:
```diff
@@ -2,6 +2,15 @@
 <dict>
+    <key>UIAppFonts</key>
+         <array>
+             <string>YS Display-Bold</string>
+             <string>YS Display-Heavy</string>
+             <string>YS Display-Light</string>
+             <string>YS Display-Medium</string>
+             <string>YS Display-Regular</string>
+             <string>YS Display-Thin</string>
```
Так какие нужны-то, выяснил?)

icydtw replied:
Сегодня явно не мой вечер! Ахаха. Честно, прям стыдно за все эти ошибки.

ABashkirova replied:
На свежую голову и не спеша получилось бы лучше :)

icydtw committed:
Исправление крайне стыдных ошибок.

icydtw committed:
Исправление ещё более стыдной ошибки.

icydtw commented:
@ABashkirova, с указанным выше разобрался. У меня есть шансы защититься через Практикум или что-нибудь ещё поправить?

ABashkirova commented:
Фон YP Black и присылай на проверку @icydtw

icydtw committed:
Изменение цвета фона.

icydtw commented:
Отправил! 🙌🏻
